diff --git a/src/node-haste/DependencyGraph.js b/src/node-haste/DependencyGraph.js
index 99f206667a7ee7075cf032cf8228135a211b36a1..e491c491ba9f9cd1d94d446ef8ad7dd556e3116f 100644
--- a/src/node-haste/DependencyGraph.js
+++ b/src/node-haste/DependencyGraph.js
@@ -177,6 +177,14 @@ class DependencyGraph extends EventEmitter {
     return nullthrows(this._fileSystem).getAllFiles();
   }
   async getOrComputeSha1(mixedPath) {
+    if (mixedPath.includes("react-native-worklets/.worklets/")) {
+      const createHash = require("crypto").createHash;
+      return {
+        sha1: createHash("sha1")
+          .update(performance.now().toString())
+          .digest("hex"),
+      };
+    }
     const result = await this._fileSystem.getOrComputeSha1(mixedPath);
     if (!result || !result.sha1) {
       throw new Error(`Failed to get the SHA-1 for: ${mixedPath}.
