diff --git a/dist/utils/loadMetroConfig.js b/dist/utils/loadMetroConfig.js
index c0749e111d342611631c3eedc9c757936e8ff225..5f98013be1ce431f0dc201341ed441675ee37363 100644
--- a/dist/utils/loadMetroConfig.js
+++ b/dist/utils/loadMetroConfig.js
@@ -32,7 +32,8 @@ function getOverrideConfig(ctx, config) {
   return {
     resolver,
     serializer: {
-      getModulesRunBeforeMainModule: () => [
+      getModulesRunBeforeMainModule: (entryFile) => [
+        ...(config.serializer.getModulesRunBeforeMainModule(entryFile) ?? []),
         require.resolve(
           _path.default.join(
             ctx.reactNativePath,
