diff --git a/react-native-slider.podspec b/react-native-slider.podspec
index 54d77bd5e64ac01af46a8c177452b8fb6c889afa..e1903fbdcb4af2a4f89ee94f0e67d59461a98eea 100644
--- a/react-native-slider.podspec
+++ b/react-native-slider.podspec
@@ -13,7 +13,7 @@ Pod::Spec.new do |s|
 
   s.authors      = package['author']
   s.homepage     = package['homepage']
-  s.platforms    = { :ios => "9.0", :visionos => "1.0" }
+  s.platforms    = { :ios => "10.0", :visionos => "1.0" }
 
   s.source       = { :git => "https://github.com/callstack/react-native-slider.git", :tag => "v#{s.version}" }
   s.source_files = "ios/**/*.{h,m,mm}"
