diff --git a/jest-preset.js b/jest-preset.js
index 5632cc7fe5c2feeb7d8474db42cbc64403963cb1..b6902abb053f65e24579795507f77428d997a4f1 100644
--- a/jest-preset.js
+++ b/jest-preset.js
@@ -17,9 +17,6 @@ module.exports = {
     defaultPlatform: 'ios',
     platforms: ['android', 'ios', 'native'],
   },
-  moduleNameMapper: {
-    '^react-native($|/.*)': `${path.dirname(require.resolve('react-native'))}/$1`,
-  },
   resolver: require.resolve('./jest/resolver.js'),
   transform: {
     '^.+\\.(js|ts|tsx)$': 'babel-jest',
