diff --git a/package.json b/package.json
index b30a5bacd99db50c5d09d7662f06bfbccca599a1..9915fa6b5d96243a00e16100da605e0f9596bf53 100644
--- a/package.json
+++ b/package.json
@@ -26,6 +26,7 @@
   "license": "MIT",
   "homepage": "https://shopify.github.io/flash-list/",
   "main": "dist/index.js",
+  "react-native": "src/index.ts",
   "types": "dist/index.d.ts",
   "scripts": {
     "up": "bundle install && yarn fixture:rn:up && yarn e2e:up && yarn build",
diff --git a/src/FlashList.tsx b/src/FlashList.tsx
index 0e9f071a53d113e85f3099ccf239d1f3d06361b6..877896f778b4fa27977b1aed9b550e2880c11e79 100644
--- a/src/FlashList.tsx
+++ b/src/FlashList.tsx
@@ -4,6 +4,7 @@ import {
   RefreshControl,
   LayoutChangeEvent,
   NativeSyntheticEvent,
+  ScrollViewComponent,
   StyleSheet,
   NativeScrollEvent,
 } from "react-native";
@@ -842,6 +843,12 @@ class FlashList<T> extends React.PureComponent<
     return this.rlvRef?.getScrollableNode?.() || null;
   }
 
+  public getNativeScrollRef(): React.ElementRef<
+    typeof ScrollViewComponent
+  > | null {
+    return this.rlvRef?.getNativeScrollRef?.() || null;
+  }
+
   /**
    * Allows access to internal recyclerlistview. This is useful for enabling access to its public APIs.
    * Warning: We may swap recyclerlistview for something else in the future. Use with caution.
diff --git a/android/src/paper/java/com/facebook/react/viewmanagers/AutoLayoutViewManagerDelegate.java b/android/src/paper/java/com/facebook/react/viewmanagers/AutoLayoutViewManagerDelegate.java
index 4c9080790230cc75f48e4ee68c84df1080986bd6..29f81e03caed5c97401de153c53fe71c3cce9e10 100644
--- a/android/src/paper/java/com/facebook/react/viewmanagers/AutoLayoutViewManagerDelegate.java
+++ b/android/src/paper/java/com/facebook/react/viewmanagers/AutoLayoutViewManagerDelegate.java
@@ -11,10 +11,11 @@ package com.facebook.react.viewmanagers;
 
 import android.view.View;
 import androidx.annotation.Nullable;
+import com.facebook.react.uimanager.BaseViewManager;
 import com.facebook.react.uimanager.BaseViewManagerDelegate;
-import com.facebook.react.uimanager.BaseViewManagerInterface;
+import com.facebook.react.uimanager.LayoutShadowNode;
 
-public class AutoLayoutViewManagerDelegate<T extends View, U extends BaseViewManagerInterface<T> & AutoLayoutViewManagerInterface<T>> extends BaseViewManagerDelegate<T, U> {
+public class AutoLayoutViewManagerDelegate<T extends View, U extends BaseViewManager<T, ? extends LayoutShadowNode> & AutoLayoutViewManagerInterface<T>> extends BaseViewManagerDelegate<T, U> {
   public AutoLayoutViewManagerDelegate(U viewManager) {
     super(viewManager);
   }
diff --git a/android/src/paper/java/com/facebook/react/viewmanagers/CellContainerManagerDelegate.java b/android/src/paper/java/com/facebook/react/viewmanagers/CellContainerManagerDelegate.java
index 2b64af7d5ff936793bf027cba249f31777daba9f..4ba6e5fd815aeac1da56374c8c48a9fcd7f57061 100644
--- a/android/src/paper/java/com/facebook/react/viewmanagers/CellContainerManagerDelegate.java
+++ b/android/src/paper/java/com/facebook/react/viewmanagers/CellContainerManagerDelegate.java
@@ -11,10 +11,11 @@ package com.facebook.react.viewmanagers;
 
 import android.view.View;
 import androidx.annotation.Nullable;
+import com.facebook.react.uimanager.BaseViewManager;
 import com.facebook.react.uimanager.BaseViewManagerDelegate;
-import com.facebook.react.uimanager.BaseViewManagerInterface;
+import com.facebook.react.uimanager.LayoutShadowNode;
 
-public class CellContainerManagerDelegate<T extends View, U extends BaseViewManagerInterface<T> & CellContainerManagerInterface<T>> extends BaseViewManagerDelegate<T, U> {
+public class CellContainerManagerDelegate<T extends View, U extends BaseViewManager<T, ? extends LayoutShadowNode> & CellContainerManagerInterface<T>> extends BaseViewManagerDelegate<T, U> {
   public CellContainerManagerDelegate(U viewManager) {
     super(viewManager);
   }
