diff --git a/src/node-haste/DependencyGraph.js b/src/node-haste/DependencyGraph.js
index 8222d8c69059bfb3fe444acba9f96ab2cc7bec0b..c4816d0c1ea52bd5965b6038ff1a159dd789cfd8 100644
--- a/src/node-haste/DependencyGraph.js
+++ b/src/node-haste/DependencyGraph.js
@@ -32,6 +32,12 @@ function getOrCreateMap(map, field) {
   }
   return subMap;
 }
+
+const workletsDirPath = _path.default.join(
+  "react-native-worklets",
+  ".worklets",
+);
+
 class DependencyGraph extends _events.default {
   #packageCache;
   #dependencyPlugin;
@@ -197,6 +203,14 @@ class DependencyGraph extends _events.default {
     return (0, _nullthrows.default)(this._fileSystem).getAllFiles();
   }
   async getOrComputeSha1(mixedPath) {
+    if (mixedPath.includes(workletsDirPath)) {
+      const createHash = require("crypto").createHash;
+      return {
+        sha1: createHash("sha1")
+          .update(performance.now().toString())
+          .digest("hex"),
+      };
+    }
     const result = await this._fileSystem.getOrComputeSha1(mixedPath);
     if (!result || !result.sha1) {
       throw new Error(`Failed to get the SHA-1 for: ${mixedPath}.
