diff --git a/src/modules/HMRClient.js b/src/modules/HMRClient.js
index 3e2652d7a43ff0d2ab3ad7556ecf79f88f7a6edb..f64ef54ac414037aa2d9c7031a7e46a0262d2e06 100644
--- a/src/modules/HMRClient.js
+++ b/src/modules/HMRClient.js
@@ -2,6 +2,12 @@
 
 const EventEmitter = require("./vendor/eventemitter3");
 const inject = ({ module: [id, code], sourceURL }) => {
+  if (
+    global._WORKLETS_BUNDLE_MODE &&
+    global.__workletsModuleProxy?.propagateModuleUpdate
+  ) {
+    global.__workletsModuleProxy.propagateModuleUpdate(code, sourceURL);
+  }
   if (global.globalEvalWithSourceUrl) {
     global.globalEvalWithSourceUrl(code, sourceURL);
   } else {
