diff --git a/src/node-haste/DependencyGraph.js b/src/node-haste/DependencyGraph.js
index bd42485a8a6647757c0f5b2f2c4a5e659125fcdd..eb4d66cd46c7347b2d0fa3dbb6c9891a71640675 100644
--- a/src/node-haste/DependencyGraph.js
+++ b/src/node-haste/DependencyGraph.js
@@ -186,6 +186,18 @@ class DependencyGraph extends _events.default {
     return (0, _nullthrows.default)(this._fileSystem).getAllFiles();
   }
   async getOrComputeSha1(mixedPath) {
+    const workletsDirPath = _path.default.join(
+      "react-native-worklets",
+      ".worklets",
+    );
+    if (mixedPath.includes(workletsDirPath)) {
+      const createHash = require("crypto").createHash;
+      return {
+        sha1: createHash("sha1")
+          .update(performance.now().toString())
+          .digest("hex"),
+      };
+    }
     const result = await this._fileSystem.getOrComputeSha1(mixedPath);
     if (!result || !result.sha1) {
       throw new Error(`Failed to get the SHA-1 for: ${mixedPath}.
