diff --git a/android/fabric/src/main/java/com/swmansion/gesturehandler/ReactContextExtensions.kt b/android/fabric/src/main/java/com/swmansion/gesturehandler/ReactContextExtensions.kt
index 2e05de9bfb63998bfbaf995780964de5eab10faa..ff440de9124f5017f9cafbef38650ec56a7b0101 100644
--- a/android/fabric/src/main/java/com/swmansion/gesturehandler/ReactContextExtensions.kt
+++ b/android/fabric/src/main/java/com/swmansion/gesturehandler/ReactContextExtensions.kt
@@ -8,5 +8,5 @@ import com.facebook.react.uimanager.events.Event
 
 fun ReactContext.dispatchEvent(event: Event<*>) {
   val fabricUIManager = UIManagerHelper.getUIManager(this, UIManagerType.FABRIC) as FabricUIManager
-  fabricUIManager.eventDispatcher.dispatchEvent(event)
+  fabricUIManager.getEventDispatcher().dispatchEvent(event)
 }
diff --git a/android/paper/src/main/java/com/swmansion/gesturehandler/ReactContextExtensions.kt b/android/paper/src/main/java/com/swmansion/gesturehandler/ReactContextExtensions.kt
index 7861838735cebcb225fb657ea4bfc4ba0a7f0a16..574dcfc59ecd86944d74f7066b2b93303dcb6c51 100644
--- a/android/paper/src/main/java/com/swmansion/gesturehandler/ReactContextExtensions.kt
+++ b/android/paper/src/main/java/com/swmansion/gesturehandler/ReactContextExtensions.kt
@@ -6,7 +6,7 @@ import com.facebook.react.uimanager.events.Event
 
 fun ReactContext.dispatchEvent(event: Event<*>) {
   try {
-    this.getNativeModule(UIManagerModule::class.java)!!.eventDispatcher.dispatchEvent(event)
+    this.getNativeModule(UIManagerModule::class.java)!!.getEventDispatcher().dispatchEvent(event)
   } catch (e: NullPointerException) {
     throw Exception("Couldn't get an instance of UIManagerModule. Gesture Handler is unable to send an event.", e)
   }
diff --git a/android/src/main/jni/CMakeLists.txt b/android/src/main/jni/CMakeLists.txt
index 74dd773ab3645cf41ab876c3b8b84bde74ff8793..5a5ecaf0633c0d5e95dadb6060ea6378d7fc32eb 100644
--- a/android/src/main/jni/CMakeLists.txt
+++ b/android/src/main/jni/CMakeLists.txt
@@ -26,12 +26,21 @@ target_include_directories(
 )
 
 find_package(ReactAndroid REQUIRED CONFIG)
-
-target_link_libraries(
-  gesturehandler
-  ReactAndroid::react_render_core
-  ReactAndroid::react_render_uimanager
-  ReactAndroid::react_render_graphics
-  ReactAndroid::jsi
-  ReactAndroid::react_nativemodule_core
-)
+if (ReactAndroid_VERSION_MINOR GREATER_EQUAL 76)
+  target_link_libraries(
+    ${PACKAGE_NAME}
+    ReactAndroid::reactnative
+    ReactAndroid::jsi
+  )
+elseif (ReactAndroid_VERSION_MINOR GREATER_EQUAL 75)
+  target_link_libraries(
+    ${PACKAGE_NAME}
+    ReactAndroid::react_render_core
+    ReactAndroid::react_render_uimanager
+    ReactAndroid::react_render_graphics
+    ReactAndroid::jsi
+    ReactAndroid::react_nativemodule_core
+  )
+else ()
+  message(FATAL_ERROR "react-native-gesture-handler on the New Architecture requires react-native 0.75 or newer.")
+endif ()
