diff --git a/android/src/paper/java/com/reactnativepagerview/PagerViewViewManager.kt b/android/src/paper/java/com/reactnativepagerview/PagerViewViewManager.kt
index 959362c62af06f791088c416e2db8f1629ab63ce..8d8da1fcfc58a8dd76ec5ccc42a1cb22ea603290 100644
--- a/android/src/paper/java/com/reactnativepagerview/PagerViewViewManager.kt
+++ b/android/src/paper/java/com/reactnativepagerview/PagerViewViewManager.kt
@@ -34,7 +34,7 @@ class PagerViewViewManager : ViewGroupManager<NestedScrollableHost>() {
         vp.adapter = ViewPagerAdapter()
         //https://github.com/callstack/react-native-viewpager/issues/183
         vp.isSaveEnabled = false
-        eventDispatcher = reactContext.getNativeModule(UIManagerModule::class.java)!!.eventDispatcher
+        eventDispatcher = reactContext.getNativeModule(UIManagerModule::class.java)!!.getEventDispatcher()
 
         vp.post {
             vp.registerOnPageChangeCallback(object : OnPageChangeCallback() {
