diff --git a/android/src/fabric/java/com/reactnativepagerview/PagerViewViewManager.kt b/android/src/fabric/java/com/reactnativepagerview/PagerViewViewManager.kt
index aa974a6bad63f5015a75834f9740abacd0771e0d..e9c4a52d4b98550b4c8f326ccf9b9bbf66cb615b 100644
--- a/android/src/fabric/java/com/reactnativepagerview/PagerViewViewManager.kt
+++ b/android/src/fabric/java/com/reactnativepagerview/PagerViewViewManager.kt
@@ -88,7 +88,7 @@ class PagerViewViewManager : ViewGroupManager<NestedScrollableHost>(), RNCViewPa
         return host
     }
 
-    override fun addView(host: NestedScrollableHost, child: View?, index: Int) {
+    override fun addView(host: NestedScrollableHost, child: View, index: Int) {
         PagerViewViewManagerImpl.addView(host, child, index)
     }
 
diff --git a/android/src/paper/java/com/reactnativepagerview/PagerViewViewManager.kt b/android/src/paper/java/com/reactnativepagerview/PagerViewViewManager.kt
index 4d0adffe6a658de709319e0700b35f726783dca2..959362c62af06f791088c416e2db8f1629ab63ce 100644
--- a/android/src/paper/java/com/reactnativepagerview/PagerViewViewManager.kt
+++ b/android/src/paper/java/com/reactnativepagerview/PagerViewViewManager.kt
@@ -69,7 +69,7 @@ class PagerViewViewManager : ViewGroupManager<NestedScrollableHost>() {
         return host
     }
 
-    override fun addView(host: NestedScrollableHost, child: View?, index: Int) {
+    override fun addView(host: NestedScrollableHost, child: View, index: Int) {
         PagerViewViewManagerImpl.addView(host, child, index)
     }
 
diff --git a/lib/typescript/PagerView.d.ts b/lib/typescript/PagerView.d.ts
index f800118e8a8132a760cd7e21c712cf9c55ace7df..003900a2bef8fc81d04d5d059e445bbe18c7b022 100644
--- a/lib/typescript/PagerView.d.ts
+++ b/lib/typescript/PagerView.d.ts
@@ -69,7 +69,7 @@ declare class PagerViewInternal extends React.Component<NativeProps> {
     render(): JSX.Element;
 }
 export declare const PagerView: React.ForwardRefExoticComponent<Omit<NativeProps, "useLegacy"> & {
-    useNext: boolean;
+    useNext?: boolean;
 } & React.RefAttributes<PagerViewInternal>>;
 export type PagerView = PagerViewInternal & typeof PagerView;
 export {};
diff --git a/src/PagerView.tsx b/src/PagerView.tsx
index 62faa7490cebb6ec8fcbaca16a35a6f4bfbad00d..79e36d5020b534db7aa869c7ddc7cb6f401125e0 100644
--- a/src/PagerView.tsx
+++ b/src/PagerView.tsx
@@ -225,7 +225,7 @@ class PagerViewInternal extends React.Component<NativeProps> {
 }
 
 // Temporary solution. It should be removed once all things get fixed
-type PagerViewProps = Omit<NativeProps, 'useLegacy'> & { useNext: boolean };
+type PagerViewProps = Omit<NativeProps, 'useLegacy'> & { useNext?: boolean };
 
 export const PagerView = React.forwardRef<PagerViewInternal, PagerViewProps>(
   (props, ref) => {
