diff --git a/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm b/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
index 74f8178a029e8706a655dd21c1c03ad76040648b..55f7f81ff2c385c89690e061391913ddcbd38019 100644
--- a/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
+++ b/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
@@ -480,6 +480,11 @@ - (void)_loadJSBundle:(NSURL *)sourceURL
         // DevSettings module is needed by _loadScriptFromSource's callback so prior initialization is required
         RCTDevSettings *const devSettings =
             (RCTDevSettings *)[strongSelf->_turboModuleManager moduleForName:"DevSettings"];
+        id worklets = [strongSelf->_turboModuleManager moduleForName:"WorkletsModule"];
+        NSString *paramMethodName = @"setBundleString:";
+        SEL paramSelector = NSSelectorFromString(paramMethodName);
+        NSData* clonedData = [source.data copy];
+        [worklets performSelector:paramSelector withObject:clonedData];
         [strongSelf _loadScriptFromSource:source];
         // Set up hot module reloading in Dev only.
         [strongSelf->_performanceLogger markStopForTag:RCTPLScriptDownload];
