diff --git a/common/cpp/react/renderer/components/rnsvg/RNSVGLayoutableShadowNode.cpp b/common/cpp/react/renderer/components/rnsvg/RNSVGLayoutableShadowNode.cpp
index a42fde95cde727639c0264b0ecdbbb7c4d2e9d53..7c66b93ccc9aa5639c81f6185332a59890eb1ef5 100644
--- a/common/cpp/react/renderer/components/rnsvg/RNSVGLayoutableShadowNode.cpp
+++ b/common/cpp/react/renderer/components/rnsvg/RNSVGLayoutableShadowNode.cpp
@@ -24,7 +24,10 @@ void RNSVGLayoutableShadowNode::setZeroDimensions() {
   // views in the layout inspector when Yoga attempts to interpret SVG
   // properties like width when viewBox scale is set.
   auto style = yogaNode_.style();
-#if REACT_NATIVE_MINOR_VERSION >= 77
+#if REACT_NATIVE_MINOR_VERSION >= 78
+  style.setDimension(yoga::Dimension::Width, yoga::StyleSizeLength::points(0));
+  style.setDimension(yoga::Dimension::Height, yoga::StyleSizeLength::points(0));
+#elif REACT_NATIVE_MINOR_VERSION >= 77
   style.setDimension(yoga::Dimension::Width, yoga::StyleLength::points(0));
   style.setDimension(yoga::Dimension::Height, yoga::StyleLength::points(0));
 #else
