diff --git a/common/cpp/react/renderer/components/rnsvg/RNSVGConcreteShadowNode.h b/common/cpp/react/renderer/components/rnsvg/RNSVGConcreteShadowNode.h
index afbd565da731c8c071a57ff1985bbccfea6f72f5..0ab02986c6de92a71ccd56908eb7c3f585621ca2 100644
--- a/common/cpp/react/renderer/components/rnsvg/RNSVGConcreteShadowNode.h
+++ b/common/cpp/react/renderer/components/rnsvg/RNSVGConcreteShadowNode.h
@@ -13,16 +13,14 @@ class RNSVGConcreteShadowNode : public ConcreteShadowNode<
                                     RNSVGLayoutableShadowNode,
                                     PropsT,
                                     ViewEventEmitter,
-                                    StateData,
-                                    false> {
+                                    StateData> {
  public:
   using BaseShadowNode = ConcreteShadowNode<
       concreteComponentName,
       RNSVGLayoutableShadowNode,
       PropsT,
       ViewEventEmitter,
-      StateData,
-      false>;
+      StateData>;
 
   using ConcreteViewProps = PropsT;
 
