diff --git a/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm b/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
index e76e1d18df99468f3cd2c9f61b20ad829081ceb5..d5ef1f4c71c0440b1e0e16e947bca06dc0491cea 100644
--- a/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
+++ b/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
@@ -512,6 +512,13 @@ - (void)_loadJSBundle:(NSURL *)sourceURL
         // DevSettings module is needed by _loadScriptFromSource's callback so prior initialization is required
         RCTDevSettings *const devSettings =
             (RCTDevSettings *)[strongSelf->_turboModuleManager moduleForName:"DevSettings"];
+
+        id worklets = [strongSelf->_turboModuleManager  moduleForName:"WorkletsModule"];
+        NSString *paramMethodName = @"setBundleString:";
+        SEL paramSelector = NSSelectorFromString(paramMethodName);
+        NSData* clonedData = [source.data copy];
+        [worklets performSelector:paramSelector withObject:clonedData];
+
         [strongSelf _loadScriptFromSource:source];
         // Set up hot module reloading in Dev only.
         [strongSelf->_performanceLogger markStopForTag:RCTPLScriptDownload];
