diff --git a/android/src/main/jni/CMakeLists.txt b/android/src/main/jni/CMakeLists.txt
index 06a3ad351aaa973d72d3a98b8847c1d299580240..f1a0539cb8cce29d419f261bd0821cdbdcd7a760 100644
--- a/android/src/main/jni/CMakeLists.txt
+++ b/android/src/main/jni/CMakeLists.txt
@@ -20,6 +20,7 @@ add_compile_options(
   -Wpedantic
   -Wno-gnu-zero-variadic-macro-arguments
   -Wno-dollar-in-identifier-extension
+  -DRN_SERIALIZABLE_STATE
 )

 file(GLOB rnsvg_SRCS CONFIGURE_DEPENDS *.cpp ${RNS_COMMON_DIR}/react/renderer/components/rnsvg/*.cpp)