diff --git a/apps/fabric-example/android/gradle.properties b/apps/fabric-example/android/gradle.properties
index 14883eb1f1..b4ffa2a4d8 100644
--- a/apps/fabric-example/android/gradle.properties
+++ b/apps/fabric-example/android/gradle.properties
@@ -50,7 +50,7 @@ enableReanimatedProfiling=true
 enableWorkletsProfiling=true
 
 # Uncomment the next line to enable bundle mode.
-# workletsBundleMode=true
+workletsBundleMode=true
 
 # https://docs.gradle.org/current/userguide/configuration_cache.html
 org.gradle.configuration-cache=true
diff --git a/apps/fabric-example/babel.config.js b/apps/fabric-example/babel.config.js
index 3e9952b8c8..a565523f7a 100644
--- a/apps/fabric-example/babel.config.js
+++ b/apps/fabric-example/babel.config.js
@@ -1,7 +1,8 @@
 /** @type {import('react-native-worklets/plugin').PluginOptions} */
 const workletsPluginOptions = {
   // Uncomment the next line to enable bundle mode.
-  // bundleMode: true,
+  bundleMode: true,
+  workletizableModules: ['axios'],
 };
 
 /** @type {import('@babel/core').TransformOptions} */
diff --git a/apps/fabric-example/ios/Podfile b/apps/fabric-example/ios/Podfile
index 586a04355d..cfb49acaf8 100644
--- a/apps/fabric-example/ios/Podfile
+++ b/apps/fabric-example/ios/Podfile
@@ -12,7 +12,7 @@ ENV['IS_REANIMATED_PROFILING'] = '1'
 ENV['IS_WORKLETS_PROFILING'] = '1'
 
 # Uncomment the next line to enable bundle mode.
-# ENV['WORKLETS_BUNDLE_MODE'] = '1'
+ENV['WORKLETS_BUNDLE_MODE'] = '1'
 
 platform :ios, min_ios_version_supported
 prepare_react_native_project!
diff --git a/apps/fabric-example/metro.config.js b/apps/fabric-example/metro.config.js
index 0d317aee07..a4953192ed 100644
--- a/apps/fabric-example/metro.config.js
+++ b/apps/fabric-example/metro.config.js
@@ -34,8 +34,8 @@ let config = {
 config = mergeConfig(defaultConfig, config);
 
 // Uncomment the following to enable bundle mode.
-// const { bundleModeMetroConfig } = require('react-native-worklets/bundleMode');
-// config = mergeConfig(config, bundleModeMetroConfig);
+const { bundleModeMetroConfig } = require('react-native-worklets/bundleMode');
+config = mergeConfig(config, bundleModeMetroConfig);
 
 module.exports = wrapWithReanimatedMetroConfig(
   mergeConfig(defaultConfig, config)
diff --git a/package.json b/package.json
index 141335c036..e4e01e00a6 100644
--- a/package.json
+++ b/package.json
@@ -70,5 +70,9 @@
     "react-native-builder-bob": "0.40.13",
     "shelljs": "0.10.0",
     "typescript": "5.8.3"
+  },
+  "resolutions": {
+    "metro": "patch:metro@npm%3A0.83.2#~/.yarn/patches/metro-npm-0.83.2-d09f48ca84.patch",
+    "metro-runtime": "patch:metro-runtime@npm%3A0.83.2#~/.yarn/patches/metro-runtime-npm-0.83.2-c614bbd3b9.patch"
   }
 }
