diff --git a/apps/fabric-example/babel.config.js b/apps/fabric-example/babel.config.js
index ea4ad7a5f8..52318bd40c 100644
--- a/apps/fabric-example/babel.config.js
+++ b/apps/fabric-example/babel.config.js
@@ -2,8 +2,8 @@
 const workletsPluginOptions = {
   strictGlobal: true,
   // Uncomment the following to enable bundle mode.
-  // bundleMode: true,
-  // workletizableModules: ['axios'],
+  bundleMode: true,
+  workletizableModules: ['axios'],
 };
 
 /** @type {import('@babel/core').TransformOptions} */
diff --git a/apps/fabric-example/metro.config.js b/apps/fabric-example/metro.config.js
index afe4c07b49..eece71ae7f 100644
--- a/apps/fabric-example/metro.config.js
+++ b/apps/fabric-example/metro.config.js
@@ -33,8 +33,8 @@ let config = {
 config = mergeConfig(defaultConfig, config);
 
 // Uncomment the following to enable bundle mode.
-// const { bundleModeMetroConfig } = require('react-native-worklets/bundleMode');
-// config = mergeConfig(config, bundleModeMetroConfig);
+const { bundleModeMetroConfig } = require('react-native-worklets/bundleMode');
+config = mergeConfig(config, bundleModeMetroConfig);
 
 module.exports = wrapWithReanimatedMetroConfig(
   mergeConfig(defaultConfig, config)
diff --git a/apps/fabric-example/package.json b/apps/fabric-example/package.json
index 85d2dbe679..d3a3b79d2a 100644
--- a/apps/fabric-example/package.json
+++ b/apps/fabric-example/package.json
@@ -45,8 +45,8 @@
     "staticFeatureFlags": {
       "RUNTIME_TEST_FLAG": true,
       "IOS_DYNAMIC_FRAMERATE_ENABLED": true,
-      "BUNDLE_MODE_ENABLED": false,
-      "FETCH_PREVIEW_ENABLED": false
+      "BUNDLE_MODE_ENABLED": true,
+      "FETCH_PREVIEW_ENABLED": true
     }
   }
 }
diff --git a/package.json b/package.json
index 390419dccc..04b68c48d4 100644
--- a/package.json
+++ b/package.json
@@ -76,5 +76,9 @@
     "react-native-builder-bob": "0.40.13",
     "shelljs": "0.10.0",
     "typescript": "5.8.3"
+  },
+  "resolutions": {
+    "metro": "patch:metro@npm%3A0.83.2#~/.yarn/patches/metro-npm-0.83.2-d09f48ca84.patch",
+    "metro-runtime": "patch:metro-runtime@npm%3A0.83.2#~/.yarn/patches/metro-runtime-npm-0.83.2-c614bbd3b9.patch"
   }
 }
