diff --git a/apps/fabric-example/babel.config.js b/apps/fabric-example/babel.config.js
index 32037220c7..a45429269f 100644
--- a/apps/fabric-example/babel.config.js
+++ b/apps/fabric-example/babel.config.js
@@ -1,8 +1,8 @@
 /** @type {import('react-native-worklets/plugin').PluginOptions} */
 const workletsPluginOptions = {
   // Uncomment the following to enable bundle mode.
-  // bundleMode: true,
-  // workletizableModules: ['axios'],
+  bundleMode: true,
+  workletizableModules: ['axios'],
 };
 
 /** @type {import('@babel/core').TransformOptions} */
diff --git a/apps/fabric-example/metro.config.js b/apps/fabric-example/metro.config.js
index 0d317aee07..a4953192ed 100644
--- a/apps/fabric-example/metro.config.js
+++ b/apps/fabric-example/metro.config.js
@@ -34,8 +34,8 @@ let config = {
 config = mergeConfig(defaultConfig, config);
 
 // Uncomment the following to enable bundle mode.
-// const { bundleModeMetroConfig } = require('react-native-worklets/bundleMode');
-// config = mergeConfig(config, bundleModeMetroConfig);
+const { bundleModeMetroConfig } = require('react-native-worklets/bundleMode');
+config = mergeConfig(config, bundleModeMetroConfig);
 
 module.exports = wrapWithReanimatedMetroConfig(
   mergeConfig(defaultConfig, config)
diff --git a/apps/fabric-example/package.json b/apps/fabric-example/package.json
index ae8474192b..4021932d9f 100644
--- a/apps/fabric-example/package.json
+++ b/apps/fabric-example/package.json
@@ -46,8 +46,8 @@
     "staticFeatureFlags": {
       "RUNTIME_TEST_FLAG": true,
       "IOS_DYNAMIC_FRAMERATE_ENABLED": true,
-      "BUNDLE_MODE_ENABLED": false,
-      "FETCH_PREVIEW_ENABLED": false
+      "BUNDLE_MODE_ENABLED": true,
+      "FETCH_PREVIEW_ENABLED": true
     }
   }
 }
diff --git a/package.json b/package.json
index 141335c036..e4e01e00a6 100644
--- a/package.json
+++ b/package.json
@@ -70,5 +70,9 @@
     "react-native-builder-bob": "0.40.13",
     "shelljs": "0.10.0",
     "typescript": "5.8.3"
+  },
+  "resolutions": {
+    "metro": "patch:metro@npm%3A0.83.2#~/.yarn/patches/metro-npm-0.83.2-d09f48ca84.patch",
+    "metro-runtime": "patch:metro-runtime@npm%3A0.83.2#~/.yarn/patches/metro-runtime-npm-0.83.2-c614bbd3b9.patch"
   }
 }
