diff --git a/apps/fabric-example/babel.config.js b/apps/fabric-example/babel.config.js
index ec37dacc04..1782f028e2 100644
--- a/apps/fabric-example/babel.config.js
+++ b/apps/fabric-example/babel.config.js
@@ -1,6 +1,6 @@
 /** @type {import('react-native-worklets/plugin').PluginOptions} */
 const workletsPluginOptions = {
-  bundleMode: false,
+  bundleMode: true,
   strictGlobal: true,
   hermesBytecode: false,
   getHBCBinary,
