diff --git a/apps/tvos-example/metro.config.js b/apps/tvos-example/metro.config.js
index 6f13f57ab4..f9ef69c8c1 100644
--- a/apps/tvos-example/metro.config.js
+++ b/apps/tvos-example/metro.config.js
@@ -28,6 +28,9 @@ let config = {
   },
 };
 
+const { bundleModeMetroConfig } = require('react-native-worklets/bundleMode');
+config = mergeConfig(config, bundleModeMetroConfig);
+
 module.exports = wrapWithReanimatedMetroConfig(
   mergeConfig(defaultConfig, config)
 );
