diff --git a/apps/tvos-example/metro.config.js b/apps/tvos-example/metro.config.js
index 11a145633b..d787851a1b 100644
--- a/apps/tvos-example/metro.config.js
+++ b/apps/tvos-example/metro.config.js
@@ -28,6 +28,9 @@ let config = {
   },
 };
 
+const { bundleModeMetroConfig } = require('react-native-worklets/bundleMode');
+config = mergeConfig(config, bundleModeMetroConfig);
+
 /** @type {import('@react-native/metro-config').MetroConfig} */
 module.exports = wrapWithReanimatedMetroConfig(
   mergeConfig(defaultConfig, config)
