diff --git a/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewState.h b/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewState.h
index d9c4c19..b4dc98b 100644
--- a/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewState.h
+++ b/node_modules/react-native-safe-area-context/common/cpp/react/renderer/components/safeareacontext/RNCSafeAreaViewState.h
@@ -1,7 +1,7 @@
 #pragma once
 
 #include <react/renderer/components/safeareacontext/Props.h>
-#include <react/renderer/graphics/Geometry.h>
+#include <react/renderer/core/graphicsConversions.h>
 #include <vector>
 
 #ifdef ANDROID
