diff --git a/node_modules/react-native/ReactAndroid/build.gradle b/node_modules/react-native/ReactAndroid/build.gradle
index 3be8b26..40d5d28 100644
--- a/node_modules/react-native/ReactAndroid/build.gradle
+++ b/node_modules/react-native/ReactAndroid/build.gradle
@@ -440,6 +440,12 @@ afterEvaluate {
     configureCMakeDebug.dependsOn(preBuild)
     configureCMakeRelWithDebInfo.dependsOn(preBuild)
 
+    if (file("$reactNativeRootDir/android/com/facebook/react/react-native/0.70.1/react-native-0.70.1-debug.aar").exists()) {
+        tasks.findByName("configureCMakeDebug[arm64-v8a]")?.enabled = false
+        tasks.findByName("buildCMakeDebug[arm64-v8a][fabricjni,hermes-executor-debug,etc]")?.enabled = false
+        tasks.findByName("externalNativeBuildDebug")?.enabled = false
+    }
+
     publishing {
         publications {
             release(MavenPublication) {
diff --git a/node_modules/react-native/ReactAndroid/hermes-engine/build.gradle b/node_modules/react-native/ReactAndroid/hermes-engine/build.gradle
index ab79d88..10b9670 100644
--- a/node_modules/react-native/ReactAndroid/hermes-engine/build.gradle
+++ b/node_modules/react-native/ReactAndroid/hermes-engine/build.gradle
@@ -243,6 +243,18 @@ afterEvaluate {
     configureCMakeRelease.dependsOn(preBuild)
     configureCMakeMinSizeRel.dependsOn(preBuild)
 
+    if(file("$reactNativeRootDir/android/com/facebook/react/hermes-engine/0.70.1/hermes-engine-0.70.1-debug.aar").exists()) {
+        tasks.findByName("downloadHermes")?.enabled = false
+        tasks.findByName("unzipHermes")?.enabled = false
+        tasks.findByName("configureBuildForHermes")?.enabled = false
+        tasks.findByName("buildHermes")?.enabled = false
+        reactNativeArchitectures().each { architecture ->
+            tasks.findByName("configureCMakeRelease[${architecture}]")?.enabled = false
+        }
+        tasks.findByName("buildCMakeRelease[arm64-v8a][libhermes]")?.enabled = false
+        tasks.findByName("externalNativeBuildDebug")?.enabled = false
+    }
+
     publishing {
         publications {
             release(MavenPublication) {
