diff --git a/node_modules/react-native-gesture-handler/RNGestureHandler.podspec b/node_modules/react-native-gesture-handler/RNGestureHandler.podspec
index 100d320..033dcd7 100644
--- a/node_modules/react-native-gesture-handler/RNGestureHandler.podspec
+++ b/node_modules/react-native-gesture-handler/RNGestureHandler.podspec
@@ -19,7 +19,6 @@ Pod::Spec.new do |s|
   if fabric_enabled
     # folly_version must match the version used in React Native
     # See folly_version in react-native/React/FBReactNativeSpec/FBReactNativeSpec.podspec
-    folly_version = '2021.06.28.00-v2'
     folly_compiler_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -Wno-comma -Wno-shorten-64-to-32'
 
     s.pod_target_xcconfig = {
@@ -32,7 +31,7 @@ Pod::Spec.new do |s|
     s.dependency "React"
     s.dependency "React-RCTFabric" # This is for fabric component
     s.dependency "React-Codegen"
-    s.dependency "RCT-Folly", folly_version
+    s.dependency "RCT-Folly"
     s.dependency "RCTRequired"
     s.dependency "RCTTypeSafety"
     s.dependency "ReactCommon/turbomodule/core"
diff --git a/node_modules/react-native-gesture-handler/android/build.gradle b/node_modules/react-native-gesture-handler/android/build.gradle
index ad5845e..84d5163 100644
--- a/node_modules/react-native-gesture-handler/android/build.gradle
+++ b/node_modules/react-native-gesture-handler/android/build.gradle
@@ -63,6 +63,11 @@ repositories {
     mavenCentral()
 }
 
+def reactNativeArchitectures() {
+    def value = project.getProperties().get("reactNativeArchitectures")
+    return value ? value.split(",") : ["armeabi-v7a", "x86", "x86_64", "arm64-v8a"]
+}
+
 android {
     compileSdkVersion safeExtGet("compileSdkVersion", 28)
 
@@ -97,6 +102,9 @@ android {
                     targets "rngesturehandler_modules"
                 }
             }
+            ndk {
+                abiFilters (*reactNativeArchitectures())
+            }
         }
     }
 
diff --git a/node_modules/react-native-gesture-handler/android/src/main/jni/Android.mk b/node_modules/react-native-gesture-handler/android/src/main/jni/Android.mk
index 185709b..926b3e7 100644
--- a/node_modules/react-native-gesture-handler/android/src/main/jni/Android.mk
+++ b/node_modules/react-native-gesture-handler/android/src/main/jni/Android.mk
@@ -28,7 +28,7 @@ LOCAL_EXPORT_C_INCLUDES := $(LOCAL_PATH) $(GENERATED_SRC_DIR)/codegen/jni
 LOCAL_SHARED_LIBRARIES := libjsi \
     libfbjni \
     libglog \
-    libfolly_json \
+    libfolly_runtime \
     libyoga \
     libreact_nativemodule_core \
     libturbomodulejsijni \
@@ -36,7 +36,6 @@ LOCAL_SHARED_LIBRARIES := libjsi \
     libreact_render_core \
     libreact_render_graphics \
     libfabricjni \
-    libfolly_futures \
     libreact_debug \
     libreact_render_componentregistry \
     libreact_render_debug \
diff --git a/node_modules/react-native-gesture-handler/ios/RNGestureHandlerModule.mm b/node_modules/react-native-gesture-handler/ios/RNGestureHandlerModule.mm
index d3d0453..7d3e508 100644
--- a/node_modules/react-native-gesture-handler/ios/RNGestureHandlerModule.mm
+++ b/node_modules/react-native-gesture-handler/ios/RNGestureHandlerModule.mm
@@ -60,7 +60,10 @@ + (BOOL)requiresMainQueueSetup
 
 - (void)invalidate
 {
-    [_manager dropAllGestureHandlers];
+    RNGestureHandlerManager *handlerManager = _manager;
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [handlerManager dropAllGestureHandlers];
+    });
     
     _manager = nil;
     
