diff --git a/node_modules/react-native/scripts/react_native_pods.rb b/node_modules/react-native/scripts/react_native_pods.rb
index 9bc989e..04c4cc4 100644
--- a/node_modules/react-native/scripts/react_native_pods.rb
+++ b/node_modules/react-native/scripts/react_native_pods.rb
@@ -690,6 +690,47 @@ def downloadAndConfigureHermesSource(react_native_path)
   Pod::UI.puts "[Hermes] Extracting Hermes (#{hermes_tag_sha})"
   system("tar -zxf #{hermes_tarball_path} --strip-components=1 --directory #{hermes_dir}")
 
+  unless ENV['RN_ARCHITECTURES'].nil? || ENV['RN_ARCHITECTURES'].strip.empty?
+    # if architectures are specified, modify hermes build scripts
+    arch = ENV['RN_ARCHITECTURES']
+
+    File.open("#{hermes_dir}/utils/build-ios-framework.sh", 'w') do |f|
+      f.puts("#!/bin/bash")
+      f.puts(". ./utils/build-apple-framework.sh")
+      f.puts("if [ ! -d destroot/Library/Frameworks/universal/hermes.xcframework ]; then")
+      f.puts("  ios_deployment_target=$(get_ios_deployment_target)")
+
+      if arch.include? "iphoneos_arm64"
+        f.puts("  build_apple_framework \"iphoneos\" \"arm64\" \"$ios_deployment_target\"")
+      end
+
+      if (arch.include? "iphonesimulator_x86_64") && (arch.include? "iphonesimulator_arm64")
+        f.puts("  build_apple_framework \"iphonesimulator\" \"x86_64;arm64\" \"$ios_deployment_target\"")
+      elsif arch.include? "iphonesimulator_x86_64"
+        f.puts("  build_apple_framework \"iphonesimulator\" \"x86_64\" \"$ios_deployment_target\"")
+      elsif arch.include? "iphonesimulator_arm64"
+        f.puts("  build_apple_framework \"iphonesimulator\" \"arm64\" \"$ios_deployment_target\"")
+      end
+
+      if (arch.include? "iphoneos") && (arch.include? "iphonesimulator")
+        f.puts("  create_universal_framework \"iphoneos\" \"iphonesimulator\"")
+      elsif arch.include? "iphoneos"
+        f.puts("  create_universal_framework \"iphoneos\"")
+      elsif arch.include? "iphonesimulator"
+        f.puts("  create_universal_framework \"iphonesimulator\"")
+      end
+      f.puts("else")
+      f.puts("  echo \"Skipping; Clean \\\"destroot\\\" to rebuild\".")
+      f.puts("fi")
+      f.fsync
+    end
+
+    File.open("#{hermes_dir}/utils/build-mac-framework.sh", 'w') do |f|
+      f.puts("#!/bin/bash")
+      f.fsync
+    end
+  end
+
   hermesc_macos_path = "#{sdks_dir}/hermesc/macos/build_host_hermesc"
   hermesc_macos_link = "#{hermes_dir}/utils/build_host_hermesc"
   if (File.exist?(hermesc_macos_path))
diff --git a/node_modules/react-native/sdks/.hermesversion b/node_modules/react-native/sdks/.hermesversion
new file mode 100644
index 0000000..0bd8b2a
--- /dev/null
+++ b/node_modules/react-native/sdks/.hermesversion
@@ -0,0 +1 @@
+73fa47c9c033e2f1bf10ebab9ac58f02c94ffff9
