diff --git a/node_modules/@react-native/community-cli-plugin/dist/utils/loadMetroConfig.js b/node_modules/@react-native/community-cli-plugin/dist/utils/loadMetroConfig.js
index c0749e1..5f98013 100644
--- a/node_modules/@react-native/community-cli-plugin/dist/utils/loadMetroConfig.js
+++ b/node_modules/@react-native/community-cli-plugin/dist/utils/loadMetroConfig.js
@@ -32,7 +32,8 @@ function getOverrideConfig(ctx, config) {
   return {
     resolver,
     serializer: {
-      getModulesRunBeforeMainModule: () => [
+      getModulesRunBeforeMainModule: (entryFile) => [
+        ...(config.serializer.getModulesRunBeforeMainModule(entryFile) ?? []),
         require.resolve(
           _path.default.join(
             ctx.reactNativePath,
