diff --git a/Libraries/AppDelegate/RCTAppSetupUtils.mm b/Libraries/AppDelegate/RCTAppSetupUtils.mm
index 21f007dba4e..a973c3863d6 100644
--- a/Libraries/AppDelegate/RCTAppSetupUtils.mm
+++ b/Libraries/AppDelegate/RCTAppSetupUtils.mm
@@ -19,6 +19,7 @@
 #import <React/RCTHTTPRequestHandler.h>
 #import <React/RCTImageLoader.h>
 #import <React/RCTNetworking.h>
+#import <React/RCTBundleConsumer.h>
 
 // Fabric
 #import <React/RCTFabricSurface.h>
@@ -67,7 +68,9 @@ void RCTAppSetupPrepareApp(UIApplication *application, BOOL turboModuleEnabled)
           classNames = dependencyProvider ? dependencyProvider.imageDataDecoderClassNames : @[];
         } else if (protocol == @protocol(RCTURLRequestHandler)) {
           classNames = dependencyProvider ? dependencyProvider.URLRequestHandlerClassNames : @[];
-        }
+        } else if (protocol == @protocol(RCTBundleConsumer)) {
+          classNames = dependencyProvider ? dependencyProvider.bundleConsumerClassNames : @[];
+        } 
 
         NSMutableArray *modules = [NSMutableArray new];
 
diff --git a/Libraries/AppDelegate/RCTDependencyProvider.h b/Libraries/AppDelegate/RCTDependencyProvider.h
index bf2d8c91a49..2520381e849 100644
--- a/Libraries/AppDelegate/RCTDependencyProvider.h
+++ b/Libraries/AppDelegate/RCTDependencyProvider.h
@@ -20,6 +20,8 @@ NS_ASSUME_NONNULL_BEGIN
 
 - (NSArray<NSString *> *)URLRequestHandlerClassNames;
 
+- (NSArray<NSString *> *)bundleConsumerClassNames;
+
 - (NSDictionary<NSString *, Class<RCTComponentViewProtocol>> *)thirdPartyFabricComponents;
 
 - (nonnull NSDictionary<NSString *, id<RCTModuleProvider>> *)moduleProviders;
diff --git a/React/Base/RCTBundleConsumer.h b/React/Base/RCTBundleConsumer.h
new file mode 100644
index 00000000000..8f1eb972424
--- /dev/null
+++ b/React/Base/RCTBundleConsumer.h
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#import <Foundation/Foundation.h>
+#import <React/NSBigStringBuffer.h>
+
+/**
+ * Provides the interface needed to register a Bundle Consumer module.
+ */
+@protocol RCTBundleConsumer <NSObject>
+
+@property (nonatomic, strong, readwrite) NSBigStringBuffer *scriptBuffer;
+
+@property (nonatomic, strong, readwrite) NSString *sourceURL;
+
+@end
diff --git a/React/CxxBridge/NSBigStringBuffer.h b/React/CxxBridge/NSBigStringBuffer.h
new file mode 100644
index 00000000000..8643c17a5ae
--- /dev/null
+++ b/React/CxxBridge/NSBigStringBuffer.h
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#import <Foundation/Foundation.h>
+
+#ifdef __cplusplus
+#import <jsireact/JSIExecutor.h>
+#import <memory>
+
+using namespace facebook;
+using namespace facebook::react;
+#endif // __cplusplus
+
+@interface NSBigStringBuffer : NSObject
+#ifdef __cplusplus
+
+{
+  std::shared_ptr<const BigStringBuffer> _buffer;
+}
+
+- (instancetype)initWithSharedPtr:(const std::shared_ptr<const BigStringBuffer> &)buffer;
+- (const std::shared_ptr<const BigStringBuffer> &)getBuffer;
+#endif // __cplusplus
+
+@end
diff --git a/React/CxxBridge/NSBigStringBuffer.mm b/React/CxxBridge/NSBigStringBuffer.mm
new file mode 100644
index 00000000000..20ec1a93bc7
--- /dev/null
+++ b/React/CxxBridge/NSBigStringBuffer.mm
@@ -0,0 +1,23 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#import "NSBigStringBuffer.h"
+
+@implementation NSBigStringBuffer
+
+- (instancetype)initWithSharedPtr:(const std::shared_ptr<const BigStringBuffer>&)buffer {
+    if (self = [super init]) {
+        _buffer = buffer;
+    }
+    return self;
+}
+
+- (const std::shared_ptr<const BigStringBuffer>&)getBuffer {
+    return _buffer;
+}
+
+@end
diff --git a/ReactAndroid/src/main/java/com/facebook/react/ReactPackageTurboModuleManagerDelegate.java b/ReactAndroid/src/main/java/com/facebook/react/ReactPackageTurboModuleManagerDelegate.java
index cfbbd336b19..7a66673bd40 100644
--- a/ReactAndroid/src/main/java/com/facebook/react/ReactPackageTurboModuleManagerDelegate.java
+++ b/ReactAndroid/src/main/java/com/facebook/react/ReactPackageTurboModuleManagerDelegate.java
@@ -7,6 +7,7 @@
 
 package com.facebook.react;
 
+import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import com.facebook.infer.annotation.Assertions;
 import com.facebook.jni.HybridData;
@@ -167,6 +168,25 @@ public abstract class ReactPackageTurboModuleManagerDelegate extends TurboModule
     return (TurboModule) resolvedModule;
   }
 
+  @NonNull
+  @Override
+  public <TInterface> List<String> getModulesConformingToInterfaceNames(@NonNull Class<TInterface> clazz) {
+    List<String> moduleNames = new ArrayList<>();
+
+    for (final ModuleProvider moduleProvider : mModuleProviders) {
+      for (final ReactModuleInfo moduleInfo : mPackageModuleInfos.get(moduleProvider).values()) {
+        if(!moduleInfo.name().equals("WorkletsModule")) {
+          continue;
+        }
+        NativeModule module = moduleProvider.getModule(moduleInfo.name());
+        if (clazz.isInstance(module)) {
+          moduleNames.add(moduleInfo.name());
+        }
+      }
+    }
+    return moduleNames;
+  }
+
   @Override
   public boolean unstable_isModuleRegistered(String moduleName) {
     for (final ModuleProvider moduleProvider : mModuleProviders) {
diff --git a/ReactAndroid/src/main/java/com/facebook/react/bridge/BundleConsumer.kt b/ReactAndroid/src/main/java/com/facebook/react/bridge/BundleConsumer.kt
new file mode 100644
index 00000000000..a9f4656e0d9
--- /dev/null
+++ b/ReactAndroid/src/main/java/com/facebook/react/bridge/BundleConsumer.kt
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package com.facebook.react.bridge
+
+import com.facebook.react.fabric.BigStringBufferWrapper
+
+public interface BundleConsumer {
+  public fun setScriptWrapper(scriptWrapper: BigStringBufferWrapper)
+  public fun setSourceURL(sourceURL: String)
+}
diff --git a/ReactAndroid/src/main/java/com/facebook/react/fabric/BigStringBufferWrapper.kt b/ReactAndroid/src/main/java/com/facebook/react/fabric/BigStringBufferWrapper.kt
new file mode 100644
index 00000000000..1d41653271a
--- /dev/null
+++ b/ReactAndroid/src/main/java/com/facebook/react/fabric/BigStringBufferWrapper.kt
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package com.facebook.react.fabric
+
+import android.annotation.SuppressLint
+import android.content.res.AssetManager
+import com.facebook.jni.HybridData
+import com.facebook.proguard.annotations.DoNotStripAny
+
+/** TODO: Description */
+@SuppressLint("MissingNativeLoadLibrary")
+@DoNotStripAny
+public class BigStringBufferWrapper {
+
+  private val mHybridData: HybridData
+
+  public constructor(fileName: String) {
+    mHybridData = initHybridFromFile(fileName)
+  }
+
+  public constructor(assetManager: AssetManager, assetURL: String) {
+    mHybridData = initHybridFromAssets(assetManager, assetURL)
+  }
+
+  private external fun initHybridFromFile(fileName: String): HybridData
+
+  private external fun initHybridFromAssets(
+          assetManager: AssetManager,
+          assetURL: String
+  ): HybridData
+}
diff --git a/ReactAndroid/src/main/java/com/facebook/react/internal/turbomodule/core/TurboModuleManager.java b/ReactAndroid/src/main/java/com/facebook/react/internal/turbomodule/core/TurboModuleManager.java
index 0ea112baee9..c195fde5eb9 100644
--- a/ReactAndroid/src/main/java/com/facebook/react/internal/turbomodule/core/TurboModuleManager.java
+++ b/ReactAndroid/src/main/java/com/facebook/react/internal/turbomodule/core/TurboModuleManager.java
@@ -251,6 +251,10 @@ public class TurboModuleManager implements TurboModuleRegistry {
     return module;
   }
 
+  public <TInterface> List<String> getModulesConformingToInterfaceNames(Class<TInterface> clazz) {
+    return mDelegate != null ? mDelegate.getModulesConformingToInterfaceNames(clazz) : Collections.emptyList();
+  }
+
   /**
    * Given a ModuleHolder, and the TurboModule's moduleName, return the TurboModule instance.
    *
diff --git a/ReactAndroid/src/main/java/com/facebook/react/internal/turbomodule/core/TurboModuleManagerDelegate.java b/ReactAndroid/src/main/java/com/facebook/react/internal/turbomodule/core/TurboModuleManagerDelegate.java
index b6ebb6fa834..ddc89eb61b8 100644
--- a/ReactAndroid/src/main/java/com/facebook/react/internal/turbomodule/core/TurboModuleManagerDelegate.java
+++ b/ReactAndroid/src/main/java/com/facebook/react/internal/turbomodule/core/TurboModuleManagerDelegate.java
@@ -45,6 +45,8 @@ public abstract class TurboModuleManagerDelegate {
   @Nullable
   public abstract TurboModule getModule(String moduleName);
 
+  public abstract <TInterface> List<String> getModulesConformingToInterfaceNames(Class<TInterface> clazz);
+
   public abstract boolean unstable_isModuleRegistered(String moduleName);
 
   /**
diff --git a/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactInstance.java b/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactInstance.java
index a5f1686ff85..45ece71b6b3 100644
--- a/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactInstance.java
+++ b/ReactAndroid/src/main/java/com/facebook/react/runtime/ReactInstance.java
@@ -22,6 +22,7 @@ import com.facebook.react.DebugCorePackage;
 import com.facebook.react.ReactPackage;
 import com.facebook.react.ViewManagerOnDemandReactPackage;
 import com.facebook.react.bridge.Arguments;
+import com.facebook.react.bridge.BundleConsumer;
 import com.facebook.react.bridge.JSBundleLoader;
 import com.facebook.react.bridge.JSBundleLoaderDelegate;
 import com.facebook.react.bridge.JavaOnlyMap;
@@ -41,6 +42,7 @@ import com.facebook.react.bridge.queue.ReactQueueConfigurationImpl;
 import com.facebook.react.bridge.queue.ReactQueueConfigurationSpec;
 import com.facebook.react.devsupport.StackTraceHelper;
 import com.facebook.react.devsupport.interfaces.DevSupportManager;
+import com.facebook.react.fabric.BigStringBufferWrapper;
 import com.facebook.react.fabric.ComponentFactory;
 import com.facebook.react.fabric.FabricUIManager;
 import com.facebook.react.fabric.FabricUIManagerBinding;
@@ -339,6 +341,15 @@ final class ReactInstance {
     }
   }
 
+  void beforeLoad(BigStringBufferWrapper scriptWrapper, String sourceURL) {
+    Collection<String> bundleConsumers = mTurboModuleManager.getModulesConformingToInterfaceNames(BundleConsumer.class);
+    for (String name : bundleConsumers) {
+      BundleConsumer module = (BundleConsumer) mTurboModuleManager.getModule(name);
+      module.setScriptWrapper(scriptWrapper);
+      module.setSourceURL(sourceURL);
+    }
+  }
+
   public void loadJSBundle(JSBundleLoader bundleLoader) {
     Systrace.beginSection(Systrace.TRACE_TAG_REACT_JAVA_BRIDGE, "ReactInstance.loadJSBundle");
     bundleLoader.loadScript(
@@ -347,19 +358,30 @@ final class ReactInstance {
           public void loadScriptFromFile(
               String fileName, String sourceURL, boolean loadSynchronously) {
             mBridgelessReactContext.setSourceURL(sourceURL);
-            loadJSBundleFromFile(fileName, sourceURL);
+
+            var script = new BigStringBufferWrapper(fileName);
+
+            beforeLoad(script, sourceURL);
+            loadJSBundle(script, sourceURL);
           }
 
           @Override
           public void loadSplitBundleFromFile(String fileName, String sourceURL) {
-            loadJSBundleFromFile(fileName, sourceURL);
+            var script = new BigStringBufferWrapper(fileName);
+
+            beforeLoad(script, sourceURL);
+            loadJSBundle(script, sourceURL);
           }
 
           @Override
           public void loadScriptFromAssets(
               AssetManager assetManager, String assetURL, boolean loadSynchronously) {
             mBridgelessReactContext.setSourceURL(assetURL);
-            loadJSBundleFromAssets(assetManager, assetURL);
+            String sourceURL = assetURL.startsWith("assets://") ? assetURL.substring("assets://".length()) : assetURL;
+            var script = new BigStringBufferWrapper(assetManager, sourceURL);
+
+            beforeLoad(script, assetURL);
+            loadJSBundle(script, assetURL);
           }
 
           @Override
@@ -482,7 +504,7 @@ final class ReactInstance {
   @DoNotStrip
   private native void installGlobals(boolean isProfiling);
 
-  private native void loadJSBundleFromFile(String fileName, String sourceURL);
+  private native void loadJSBundle(BigStringBufferWrapper fileName, String sourceURL);
 
   private native void loadJSBundleFromAssets(AssetManager assetManager, String assetURL);
 
diff --git a/ReactAndroid/src/main/jni/react/fabric/BigStringBufferWrapper.cpp b/ReactAndroid/src/main/jni/react/fabric/BigStringBufferWrapper.cpp
new file mode 100644
index 00000000000..0efb9dac6b6
--- /dev/null
+++ b/ReactAndroid/src/main/jni/react/fabric/BigStringBufferWrapper.cpp
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "BigStringBufferWrapper.h"
+#include <cxxreact/JSBigString.h>
+#include <cxxreact/RecoverableError.h>
+
+using namespace facebook::jni;
+
+namespace facebook::react {
+jni::local_ref<BigStringBufferWrapper::jhybriddata>
+BigStringBufferWrapper::initHybridFromFile(
+    jni::alias_ref<jhybridobject> jThis,
+    std::string fileName) {
+  std::unique_ptr<const JSBigFileString> script;
+  RecoverableError::runRethrowingAsRecoverable<std::system_error>(
+      [&fileName, &script]() { script = JSBigFileString::fromPath(fileName); });
+  auto buffer = std::make_shared<BigStringBuffer>(std::move(script));
+  return makeCxxInstance(buffer);
+}
+
+jni::local_ref<BigStringBufferWrapper::jhybriddata>
+BigStringBufferWrapper::initHybridFromAssets(
+    jni::alias_ref<jhybridobject> jThis,
+    jni::alias_ref<JAssetManager::javaobject> assetManager,
+    const std::string& sourceURL) {
+  auto manager = extractAssetManager(assetManager);
+  auto script = loadScriptFromAssets(manager, sourceURL);
+  auto buffer = std::make_shared<BigStringBuffer>(std::move(script));
+  return makeCxxInstance(buffer);
+}
+
+BigStringBufferWrapper::BigStringBufferWrapper(
+    const std::shared_ptr<const BigStringBuffer>& script)
+    : script_(script) {}
+
+const std::shared_ptr<const BigStringBuffer> BigStringBufferWrapper::getScript()
+    const {
+  return script_;
+}
+
+void BigStringBufferWrapper::registerNatives() {
+  registerHybrid(
+      {makeNativeMethod(
+           "initHybridFromFile", BigStringBufferWrapper::initHybridFromFile),
+       makeNativeMethod(
+           "initHybridFromAssets",
+           BigStringBufferWrapper::initHybridFromAssets)});
+}
+} // namespace facebook::react
diff --git a/ReactAndroid/src/main/jni/react/fabric/BigStringBufferWrapper.h b/ReactAndroid/src/main/jni/react/fabric/BigStringBufferWrapper.h
new file mode 100644
index 00000000000..248f99faf1e
--- /dev/null
+++ b/ReactAndroid/src/main/jni/react/fabric/BigStringBufferWrapper.h
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include <fbjni/fbjni.h>
+#include <jsireact/JSIExecutor.h>
+#include <react/jni/JSLoader.h>
+
+namespace facebook::react {
+
+class BigStringBufferWrapper : public jni::HybridClass<BigStringBufferWrapper> {
+ public:
+  constexpr static const char* const kJavaDescriptor =
+      "Lcom/facebook/react/fabric/BigStringBufferWrapper;";
+
+  static void registerNatives();
+
+  [[nodiscard]] const std::shared_ptr<const BigStringBuffer> getScript() const;
+
+ private:
+  static jni::local_ref<BigStringBufferWrapper::jhybriddata> initHybridFromFile(
+      jni::alias_ref<jhybridobject> jThis,
+      std::string fileName);
+
+  static jni::local_ref<BigStringBufferWrapper::jhybriddata>
+  initHybridFromAssets(
+      jni::alias_ref<jhybridobject> jThis,
+      jni::alias_ref<JAssetManager::javaobject> assetManager,
+      const std::string& assetURL);
+
+  friend HybridBase;
+
+  explicit BigStringBufferWrapper(
+      const std::shared_ptr<const BigStringBuffer>& script);
+
+  const std::shared_ptr<const BigStringBuffer> script_;
+};
+
+} // namespace facebook::react
diff --git a/ReactAndroid/src/main/jni/react/fabric/OnLoad.cpp b/ReactAndroid/src/main/jni/react/fabric/OnLoad.cpp
index 064aec9141b..78a7ee32019 100644
--- a/ReactAndroid/src/main/jni/react/fabric/OnLoad.cpp
+++ b/ReactAndroid/src/main/jni/react/fabric/OnLoad.cpp
@@ -7,6 +7,7 @@
 
 #include <fbjni/fbjni.h>
 
+#include "BigStringBufferWrapper.h"
 #include "ComponentFactory.h"
 #include "EventBeatManager.h"
 #include "EventEmitterWrapper.h"
@@ -22,5 +23,6 @@ JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM* vm, void*) {
     facebook::react::StateWrapperImpl::registerNatives();
     facebook::react::ComponentFactory::registerNatives();
     facebook::react::SurfaceHandlerBinding::registerNatives();
+    facebook::react::BigStringBufferWrapper::registerNatives();
   });
 }
diff --git a/ReactAndroid/src/main/jni/react/runtime/jni/JReactInstance.cpp b/ReactAndroid/src/main/jni/react/runtime/jni/JReactInstance.cpp
index 7a0edf60c8b..e45873611b3 100644
--- a/ReactAndroid/src/main/jni/react/runtime/jni/JReactInstance.cpp
+++ b/ReactAndroid/src/main/jni/react/runtime/jni/JReactInstance.cpp
@@ -136,24 +136,11 @@ jni::local_ref<JReactInstance::jhybriddata> JReactInstance::initHybrid(
       jReactHostInspectorTarget);
 }
 
-void JReactInstance::loadJSBundleFromAssets(
-    jni::alias_ref<JAssetManager::javaobject> assetManager,
-    const std::string& assetURL) {
-  const int kAssetsLength = 9; // strlen("assets://");
-  auto sourceURL = assetURL.substr(kAssetsLength);
-
-  auto manager = extractAssetManager(assetManager);
-  auto script = loadScriptFromAssets(manager, sourceURL);
-  instance_->loadScript(std::move(script), sourceURL);
-}
-
-void JReactInstance::loadJSBundleFromFile(
-    const std::string& fileName,
+void JReactInstance::loadJSBundle(
+    jni::alias_ref<BigStringBufferWrapper::javaobject> scriptWrapper,
     const std::string& sourceURL) {
-  std::unique_ptr<const JSBigFileString> script;
-  RecoverableError::runRethrowingAsRecoverable<std::system_error>(
-      [&fileName, &script]() { script = JSBigFileString::fromPath(fileName); });
-  instance_->loadScript(std::move(script), sourceURL);
+  auto script = scriptWrapper->cthis()->getScript();
+  instance_->loadScript(script, sourceURL);
 }
 
 /**
@@ -223,10 +210,7 @@ void JReactInstance::registerNatives() {
       makeNativeMethod("initHybrid", JReactInstance::initHybrid),
       makeNativeMethod(
           "createJSTimerExecutor", JReactInstance::createJSTimerExecutor),
-      makeNativeMethod(
-          "loadJSBundleFromAssets", JReactInstance::loadJSBundleFromAssets),
-      makeNativeMethod(
-          "loadJSBundleFromFile", JReactInstance::loadJSBundleFromFile),
+      makeNativeMethod("loadJSBundle", JReactInstance::loadJSBundle),
       makeNativeMethod(
           "getJSCallInvokerHolder", JReactInstance::getJSCallInvokerHolder),
       makeNativeMethod(
diff --git a/ReactAndroid/src/main/jni/react/runtime/jni/JReactInstance.h b/ReactAndroid/src/main/jni/react/runtime/jni/JReactInstance.h
index d6552a8a978..ff9445980f7 100644
--- a/ReactAndroid/src/main/jni/react/runtime/jni/JReactInstance.h
+++ b/ReactAndroid/src/main/jni/react/runtime/jni/JReactInstance.h
@@ -13,6 +13,7 @@
 #include <fbjni/fbjni.h>
 #include <jni.h>
 #include <jsi/jsi.h>
+#include <react/fabric/BigStringBufferWrapper.h>
 #include <react/jni/JMessageQueueThread.h>
 #include <react/jni/JRuntimeExecutor.h>
 #include <react/jni/JRuntimeScheduler.h>
@@ -58,12 +59,8 @@ class JReactInstance : public jni::HybridClass<JReactInstance> {
 
   static void registerNatives();
 
-  void loadJSBundleFromAssets(
-      jni::alias_ref<JAssetManager::javaobject> assetManager,
-      const std::string& assetURL);
-
-  void loadJSBundleFromFile(
-      const std::string& fileName,
+  void loadJSBundle(
+      jni::alias_ref<BigStringBufferWrapper::javaobject> scriptWrapper,
       const std::string& sourceURL);
 
   void callFunctionOnModule(
diff --git a/ReactCommon/react/runtime/ReactInstance.cpp b/ReactCommon/react/runtime/ReactInstance.cpp
index 8c2c0840ac7..bba92247fa5 100644
--- a/ReactCommon/react/runtime/ReactInstance.cpp
+++ b/ReactCommon/react/runtime/ReactInstance.cpp
@@ -213,20 +213,24 @@ std::string simpleBasename(const std::string& path) {
  * preferably via the runtimeExecutor_.
  */
 void ReactInstance::loadScript(
-    std::unique_ptr<const JSBigString> script,
+    const std::shared_ptr<const BigStringBuffer>& script,
     const std::string& sourceURL,
-    std::function<void(jsi::Runtime& runtime)>&& completion) {
-  auto buffer = std::make_shared<BigStringBuffer>(std::move(script));
+    std::function<void(jsi::Runtime& runtime)>&& beforeLoad,
+    std::function<void(jsi::Runtime& runtime)>&& afterLoad) {
   std::string scriptName = simpleBasename(sourceURL);
 
   runtimeScheduler_->scheduleWork([this,
                                    scriptName,
                                    sourceURL,
-                                   buffer = std::move(buffer),
+                                   script,
                                    weakBufferedRuntimeExecuter =
                                        std::weak_ptr<BufferedRuntimeExecutor>(
                                            bufferedRuntimeExecutor_),
-                                   completion](jsi::Runtime& runtime) {
+                                   beforeLoad,
+                                   afterLoad](jsi::Runtime& runtime) {
+    if (beforeLoad) {
+      beforeLoad(runtime);
+    }
     TraceSection s("ReactInstance::loadScript");
     bool hasLogger(ReactMarker::logTaggedMarkerBridgelessImpl);
     if (hasLogger) {
@@ -234,7 +238,7 @@ void ReactInstance::loadScript(
           ReactMarker::RUN_JS_BUNDLE_START, scriptName.c_str());
     }
 
-    runtime.evaluateJavaScript(buffer, sourceURL);
+    runtime.evaluateJavaScript(script, sourceURL);
 
     /**
      * TODO(T183610671): We need a safe/reliable way to enable the js
@@ -255,8 +259,8 @@ void ReactInstance::loadScript(
             weakBufferedRuntimeExecuter.lock()) {
       strongBufferedRuntimeExecuter->flush();
     }
-    if (completion) {
-      completion(runtime);
+    if (afterLoad) {
+      afterLoad(runtime);
     }
   });
 }
diff --git a/ReactCommon/react/runtime/ReactInstance.h b/ReactCommon/react/runtime/ReactInstance.h
index a1e40537ecc..2868f9ef52c 100644
--- a/ReactCommon/react/runtime/ReactInstance.h
+++ b/ReactCommon/react/runtime/ReactInstance.h
@@ -48,9 +48,10 @@ class ReactInstance final : private jsinspector_modern::InstanceTargetDelegate {
       BindingsInstallFunc bindingsInstallFunc) noexcept;
 
   void loadScript(
-      std::unique_ptr<const JSBigString> script,
+      const std::shared_ptr<const BigStringBuffer>& script,
       const std::string& sourceURL,
-      std::function<void(jsi::Runtime& runtime)>&& completion = nullptr);
+      std::function<void(jsi::Runtime& runtime)>&& beforeLoad = nullptr,
+      std::function<void(jsi::Runtime& runtime)>&& afterLoad = nullptr);
 
   void registerSegment(uint32_t segmentId, const std::string& segmentPath);
 
diff --git a/ReactCommon/react/runtime/platform/ios/React-RuntimeApple.podspec b/ReactCommon/react/runtime/platform/ios/React-RuntimeApple.podspec
index 9bf22c28a88..a3e26944fe7 100644
--- a/ReactCommon/react/runtime/platform/ios/React-RuntimeApple.podspec
+++ b/ReactCommon/react/runtime/platform/ios/React-RuntimeApple.podspec
@@ -23,6 +23,7 @@ folly_dep_name = folly_config[:dep_name]
 
 boost_config = get_boost_config()
 boost_compiler_flags = boost_config[:compiler_flags]
+bundle_mode_flag = ENV["WORKLETS_BUNDLE_MODE"] == "1" ? 'WORKLETS_BUNDLE_MODE=1' : ''
 
 header_search_paths = [
   "$(PODS_ROOT)/boost",
@@ -45,9 +46,13 @@ Pod::Spec.new do |s|
   s.pod_target_xcconfig    = { "HEADER_SEARCH_PATHS" => header_search_paths,
                                 "USE_HEADERMAP" => "YES",
                                 "CLANG_CXX_LANGUAGE_STANDARD" => rct_cxx_language_standard(),
-                                "GCC_WARN_PEDANTIC" => "YES" }
+                                "GCC_WARN_PEDANTIC" => "YES",
+                                "GCC_PREPROCESSOR_DEFINITIONS[config=*Debug*]" => "$(inherited) #{bundle_mode_flag}",
+                                "GCC_PREPROCESSOR_DEFINITIONS[config=*Release*]" => "$(inherited) #{bundle_mode_flag}",
+                               }
   s.compiler_flags       = folly_compiler_flags + ' ' + boost_compiler_flags
 
+
   if ENV['USE_FRAMEWORKS']
     s.header_mappings_dir     = './'
     s.module_name             = 'React_RuntimeApple'
diff --git a/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm b/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
index 39285dcaaab..95efbbc43e3 100644
--- a/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
+++ b/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm
@@ -11,6 +11,7 @@
 
 #import <FBReactNativeSpec/FBReactNativeSpec.h>
 #import <React/NSDataBigString.h>
+#import <React/NSBigStringBuffer.h>
 #import <React/RCTAssert.h>
 #import <React/RCTBridge+Inspector.h>
 #import <React/RCTBridge+Private.h>
@@ -19,6 +20,7 @@
 #import <React/RCTBridgeModuleDecorator.h>
 #import <React/RCTBridgeProxy+Cxx.h>
 #import <React/RCTBridgeProxy.h>
+#import <React/RCTBundleConsumer.h>
 #import <React/RCTComponentViewFactory.h>
 #import <React/RCTConstants.h>
 #import <React/RCTCxxUtils.h>
@@ -42,6 +44,7 @@
 #import <react/renderer/runtimescheduler/RuntimeSchedulerCallInvoker.h>
 #import <react/utils/ContextContainer.h>
 #import <react/utils/ManagedObjectWrapper.h>
+#import <ReactCodegen/RCTModulesConformingToProtocolsProvider.h>
 
 #import "ObjCTimerRegistry.h"
 #import "RCTJSThreadManager.h"
@@ -494,10 +497,24 @@ - (void)_loadScriptFromSource:(RCTSource *)source
   }
 
   auto script = std::make_unique<NSDataBigString>(source.data);
+  const auto scriptBuffer = std::make_shared<const BigStringBuffer>(std::move(script));
   const auto *url = deriveSourceURL(source.url).UTF8String;
-  _reactInstance->loadScript(std::move(script), url, [](jsi::Runtime &_) {
+
+  auto beforeLoad = [turboModuleManager = self->_turboModuleManager, scriptBuffer, url](jsi::Runtime &_) {
+    #ifdef WORKLETS_BUNDLE_MODE
+    auto bundleConsumerNames = [RCTModulesConformingToProtocolsProvider bundleConsumerClassNames];
+    for (id name in bundleConsumerNames) {
+      id<RCTBundleConsumer> module = (id<RCTBundleConsumer>)[turboModuleManager moduleForName:[name UTF8String]];
+      module.scriptBuffer = [[NSBigStringBuffer alloc] initWithSharedPtr:scriptBuffer];
+      module.sourceURL = @(url);
+    }
+    #endif // WORKLETS_BUNDLE_MODE
+  };
+  auto afterLoad = [](jsi::Runtime &_) {
     [[NSNotificationCenter defaultCenter] postNotificationName:@"RCTInstanceDidLoadBundle" object:nil];
-  });
+  };
+
+  _reactInstance->loadScript(scriptBuffer, url, beforeLoad, afterLoad);
 }
 
 - (void)_handleJSError:(const JsErrorHandler::ProcessedError &)error withRuntime:(jsi::Runtime &)runtime
diff --git a/scripts/codegen/generate-artifacts-executor.js b/scripts/codegen/generate-artifacts-executor.js
index c8311663b20..7174ba86b98 100644
--- a/scripts/codegen/generate-artifacts-executor.js
+++ b/scripts/codegen/generate-artifacts-executor.js
@@ -649,11 +649,21 @@ function generateCustomURLHandlers(libraries, outputDir) {
     .map(className => `@"${className}"`)
     .join(',\n\t\t\t');
 
+  const bundleConsumerClasses = libraries
+    .flatMap(
+      library =>
+        library?.config?.ios?.modulesConformingToProtocol?.RCTBundleConsumer,
+    )
+    .filter(Boolean)
+    .map(className => `@"${className}"`)
+    .join(',\n\t\t\t');
+
   const template = fs.readFileSync(MODULES_PROTOCOLS_MM_TEMPLATE_PATH, 'utf8');
   const finalMMFile = template
     .replace(/{imageURLLoaderClassNames}/, customImageURLLoaderClasses)
     .replace(/{imageDataDecoderClassNames}/, customImageDataDecoderClasses)
-    .replace(/{requestHandlersClassNames}/, customURLHandlerClasses);
+    .replace(/{requestHandlersClassNames}/, customURLHandlerClasses)
+    .replace(/{bundleConsumerClassNames}/, bundleConsumerClasses);
 
   fs.mkdirSync(outputDir, {recursive: true});
 
diff --git a/scripts/codegen/templates/RCTModulesConformingToProtocolsProviderH.template b/scripts/codegen/templates/RCTModulesConformingToProtocolsProviderH.template
index 10eb8489175..6190c8da333 100644
--- a/scripts/codegen/templates/RCTModulesConformingToProtocolsProviderH.template
+++ b/scripts/codegen/templates/RCTModulesConformingToProtocolsProviderH.template
@@ -15,4 +15,6 @@
 
 +(NSArray<NSString *> *)URLRequestHandlerClassNames;
 
++(NSArray<NSString *> *)bundleConsumerClassNames;
+
 @end
diff --git a/scripts/codegen/templates/RCTModulesConformingToProtocolsProviderMM.template b/scripts/codegen/templates/RCTModulesConformingToProtocolsProviderMM.template
index 7a432b968ca..1e9cf8d7fe2 100644
--- a/scripts/codegen/templates/RCTModulesConformingToProtocolsProviderMM.template
+++ b/scripts/codegen/templates/RCTModulesConformingToProtocolsProviderMM.template
@@ -51,4 +51,18 @@
   return classNames;
 }
 
++(NSArray<NSString *> *)bundleConsumerClassNames
+{
+  static NSArray<NSString *> *classNames = nil;
+  static dispatch_once_t onceToken;
+  
+  dispatch_once(&onceToken, ^{
+    classNames = @[
+      {bundleConsumerClassNames}
+    ];
+  });
+  
+  return classNames;
+}
+
 @end
